src/server/location.rs

```diff
@@ -0,0 +1,356 @@
+    for lang in langs {
+        // It's a wildcard, so just return the first possible choice.
+        if lang == "*" || lang == "-" {
+            return elements.values().next().map(|v| (*v).to_string());
+        }
+        if elements.contains_key(lang.as_str()) {
+            if let Some(element) = elements.get(lang.as_str()) {
+                return Some(element.to_string());
+            }
+        }
+        if lang.contains('-') {
+            let (lang, _) = lang.split_at(2);
+            if elements.contains_key(lang) {
+                if let Some(element) = elements.get(lang) {
+                    return Some(element.to_string());
+                }
+            }
+        }
+    }
+    None
+}
+
+/// IP address to Location
+impl Location {
+    /// Is the location look-up service available?
+    pub fn is_available(&self) -> bool {
+        self.iploc.is_some()
+    }
+
+    /// Resolve an `ip_addr` to a `LocationResult` using the `preferred_languages` as a hint for the language to use.
+    ///
+    /// `preferred_languages` is an array of `Accepted-Langauge` type pairs. You can use `preferred_languages` to
+    /// convert the `Accepted-Language` header into this set.
+    pub async fn mmdb_locate(
+        &self,
+        ip_addr: IpAddr,
+        preferred_languages: &[String],
+    ) -> HandlerResult<Option<LocationResult>> {
+        if self.iploc.is_none() {
+            return Ok(None);
+        }
+        /*
+        The structure of the returned maxminddb free record is:
+        City:maxminddb::geoip::model::City {
+            city: Some(City{
+                geoname_id: Some(#),
+                names: Some({"lang": "name", ...})
+                }),
+            continent: Some(Continent{
+                geoname_id: Some(#),
+                names: Some({...})
+                }),
+            country: Some(Country{
+                geoname_id: Some(#),
+                names: Some({...})
+                }),
+            location: Some(Location{
+                latitude: Some(#.#),
+                longitude: Some(#.#),
+                metro_code: Some(#),
+                time_zone: Some(".."),
+                }),
+            postal: Some(Postal {
+                code: Some("..")
+                }),
+            registered_country: Some(Country {
+                geoname_id: Some(#),
+                iso_code: Some(".."),
+                names: Some({"lang": "name", ...})
+                }),
+            represented_country: None,
+            subdivisions: Some([Subdivision {
+                geoname_id: Some(#),
+                iso_code: Some(".."),
+                names: Some({"lang": "name", ...})
+                }]),
+            traits: None }
+        }
+        */
+        let mut result = LocationResult::default();
+        match self.iploc.clone().unwrap().lookup::<City<'_>>(ip_addr) {
+            Ok(location) => {
+                if let Some(names) = location.city.and_then(|c| c.names) {
```

Comment:
Note: I don't see us needing the language handling methods, we'd pass the country/most specific subdivision `iso_code` instead of one of `names` (though we may need to look into City's `names` if we begin using it for large enough cities).

Reply:
Yeah, the language bit is more a relic I brought over from the connection server code, mostly since MaxMind returns a set of names sorted in alphabetical by location. We *could* absolutely just force to `en` here.